fmt/src/git.rs

```diff
@@ -11,48 +11,150 @@
             None => {
                 let message = "bare repository detected";
-                if is_auto {
-                    info!("git.ignore=auto is resolved to fallback; {message}");
-                    Ok(None)
+                if feature.is_auto() {
+                    info!(?config, "git config is resolved to disabled; {message}");
+                    Ok(GitContext { repo: None, config })
                 } else {
                     InvalidConfigSnafu { message }.fail()
                 }
             }
             Some(_) => {
-                info!("git.ignore=auto is resolved to enabled");
-                Ok(Some(repo))
+                info!("git config is resolved to enabled");
+                Ok(GitContext {
+                    repo: Some(repo),
+                    config,
+                })
             }
         },
         Err(err) => {
-            if is_auto {
-                info!(?err, "git.ignore=auto is resolved to disabled");
-                Ok(None)
+            if feature.is_auto() {
+                info!(?err, ?config, "git config is resolved to disabled");
+                Ok(GitContext { repo: None, config })
             } else {
                 Err(GixDiscoverOpSnafu {}.into_error(Box::new(err)))
             }
         }
     }
 }
+
+fn resolve_features(config: &config::Git) -> FeatureGate {
+    let features = [config.attrs, config.ignore];
+    for feature in features.iter() {
+        if feature.is_enable() {
+            return FeatureGate::Enable;
+        }
+    }
+    for feature in features.iter() {
+        if feature.is_auto() {
+            return FeatureGate::Auto;
+        }
+    }
+    FeatureGate::Disable
+}
+
+#[derive(Debug)]
+pub struct GitFileAttrs {
+    pub created_time: gix::date::Time,
+    pub modified_time: gix::date::Time,
+}
+
+pub fn resolve_file_attrs(
+    git_context: GitContext,
+) -> anyhow::Result<HashMap<String, GitFileAttrs>> {
+    let mut attrs = HashMap::new();
+
+    if git_context.config.attrs.is_disable() {
+        return Ok(attrs);
+    }
+
+    let repo = match git_context.repo {
+        Some(repo) => repo,
+        None => return Ok(attrs),
+    };
+
+    let workdir = repo.work_dir().expect("workdir cannot be absent");
+    let workdir = workdir.canonicalize()?;
+
+    let mode = gix::diff::blob::pipeline::Mode::ToGit;
+    let mut cache = repo.diff_resource_cache(mode, Default::default())?;
+
+    let head = repo.head_commit()?;
+    let mut prev_commit = head.clone();
+
+    for info in head.ancestors().all()? {
+        let info = info?;
+        let this_commit = info.object()?;
+        let time = this_commit.time()?;
+
+        let tree = this_commit.tree()?;
+        let mut changes = tree.changes()?;
+        changes.track_path().for_each_to_obtain_tree_with_cache(
+            &prev_commit.tree()?,
+            &mut cache,
+            |change| {
+                let filepath = workdir.join(change.location.to_string());
```

Comment:
Here you want to use `gix::path::from_bstr(change.location)`, never use `String` when anything path related is happening as they can't represent everything that's possible on the filesystem.

Reply:
Thanks for your advice. I try to write:

```rust
                let filepath = gix::path::from_bstring(change.location);
                match attrs.entry(filepath) {
```

But the filepath is something like `"fmt/tests/tests.rs"` which cannot match the selections `"/Users/tison/Brittani/hawkeye-native/fmt/tests/tests.rs"` on the later get.

Reply:
Calling `filepath.canonicalize()` here will result in IO error `No such file or directory`. I guess it's because no base dir specified.

Perhaps I can use still `workdir.join(filepath)`?

Reply:
Git tracks paths relative to the repository root, hence one will have to join them with the worktree root before using them on the filesystem.

Reply:
Yeah. Thanks for your information.

This changes iteration currently take about 4 seconds on a repo having ~1300 files and ~2600 commits.

I wonder what is the major performance factor and whether we can improve it (e.g., does `git` work well in such situations?).

It seems most cycles are used to parse the tree data which we can do nothing to improve.

BTW, gitoxide already out performance than git command as:

```shell
for i in $(git ls-files); do git --no-pager log --follow --format=%ad -- $i > /dev/null; done
```

takes 1 minute and 5 seconds to finish.

Reply:
It's hard to get an apples-apples comparison since Git definitely does way more work here as it goes through the whole history once per file, but the Rust code only has to go through once. From that point of view, Git's performance is impressive.

> I wonder what is the major performance factor and whether we can improve it (e.g., does `git` work well in such situations?).

Object-access performance is critical, and there are some variables regarding caches that can be set. They can improve performance by a couple of percent, but it's nothing more significant. 
But wait, [in this line](https://github.com/korandoru/hawkeye/blob/main/fmt/Cargo.toml#L30) the `max-performance` feature can be added, it should be noticable.